docs/blog/20240122-savetheclocktower-modern-tree-sitter-part-6.md

```diff
@@ -0,0 +1,190 @@
+
+The good news is that the language server specification includes several actions that are relevant to `symbols-view`: [`textDocument/documentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol) for same-file symbols, [`workspace/symbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_symbol) for project-wide symbols, and even [`textDocument/definition`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_definition) for finding where a symbol is defined. Some IDE backend packages already have “brains” capable of doing these tasks!
+
+But here’s the bad news: since the `symbol.provider` service has only just been invented, those IDE backend packages need updates before they can be used for symbol navigation.
+
+I’ve started to do a bit of that work. Inspired by `ide-typescript` — but mainly starting fresh — I’ve been working on a package currently called [`pulsar-ide-typescript-alpha`](https://web.pulsar-edit.dev/packages/pulsar-ide-typescript-alpha) that aims to be its drop-in replacement. It should be able to do everything that `ide-typescript` can already do, but it will also be able to offer project-wide symbol search and go-to-declaration functionality.
+
+And it might take a few version bumps on dependencies, but most other IDE backend packages can also be updated to take advantage of these features.
+
+### Anyway, back to `symbols-view`
+
+Unlike `autocomplete-plus`, which aggregates suggestions from multiple providers and shows all of them to the user, `symbols-view` is mainly interested in choosing _the_ best provider for the job. There’s little point in aggregating across a language server _and_ Tree-sitter _and_ `ctags`, since they’re largely going to be offering the same list of symbols with varying degrees of richness, and you’d be pretty annoyed if Pulsar offered you three different list entries for the same function. Inside `symbols-view` they’re called “exclusive” providers because only one of them will be picked for the job.
+
+But I wanted to leave the door open for some creative and unexpected usages, so `symbols-view` also has a concept of “supplemental” providers. A provider that marks itself as supplemental is saying it’d like to contribute symbols that would probably not already be in an exclusive provider’s list.
+
+Did you know you can bookmark lines in a buffer? Try it out — right-click on any line of your editor and select **Toggle Bookmark**. The built-in `bookmarks` package keeps track of them and will also let you navigate between them via <kbd>F2</kbd> and <kbd>Shift+F2</kbd>. They even
```

Comment:
Aside from cleaning up the fragment at the end of this graf, I realized how abrupt this transition is, and how much it might've seemed like this paragraph belonged in another section, so I wrote a better segue.